### PR TITLE
Add large-scale Thompson sampling tutorial

### PR DESCRIPTION
**Pull request by dme65**
**Note**: This PR is work in progress so no need to review it yet, but I'm putting it up as it's relevant to #649.

This PR adds a tutorial for using large-scale Thompson sampling using Cholesky, random Fourier features (RFFs), and contour integral quadrature (CIQ). We can potentially add additional methods such as optimizing a sample path using a quasi-Newton optimizer, but that would require `RBFKernelGrad` to support incomplete gradient information.

We want to wait until #652 is landed before merging this so we can delete some of the custom TS code that I added to avoid using Cholesky in the posterior calls when using RFFs and CIQ. I'll also make some changes to be consistent with the `SingleTaskGP` priors, but for now I just added some simple bounds to the hyperparameters.

### CODE REVIEW

**Comment by eytan** [comment]
Just out of curiosity, how difficult would it be to modify `RBFKernelGrad` to support incomplete gradient information?  Apologies for the naive question here, but why would we need to modify `RBFKernelGrad` to support doing gradient descent over a RFF model?

**Comment by dme65** [comment]
> Just out of curiosity, how difficult would it be to modify `RBFKernelGrad` to support incomplete gradient information? Apologies for the naive question here, but why would we need to modify `RBFKernelGrad` to support doing gradient descent over a RFF model?

For the first question, pretty easy since it's just a matter of selecting a subset of the rows in the blocks with the first and second order derivative information. Would be nice to get that merged into GPyTorch at some point. That functionality will be useful if we want to optimize the sample path of an exact GP with a gradient-based optimizer, but it will also lead to a lot of numerical issues. It sounds like there are better alternatives here like the decoupled sampling you suggested offline.

**Comment by Balandat** [comment]
The tutorial run fails for lack of pykeops. You will want to edit the `[tutorials]` option in `setup.py` (https://github.com/pytorch/botorch/blob/master/setup.py#L34), and then also edit the CI installation scripts for pip (https://github.com/pytorch/botorch/blob/master/scripts/install_via_pip.sh) and conda (https://github.com/pytorch/botorch/blob/master/scripts/install_via_conda.sh).

File: tutorials/thompson_sampling.ipynb [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    "    return hart6(unnormalize(x, hart6.bounds))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_initial_points(dim, n_pts, seeed=None):\n",
```
Comment (saitcakmak) [inline tutorials/thompson_sampling.ipynb:90]:
```suggestion
    "def get_initial_points(dim, n_pts, seed=None):\n",
```
just a minor typo

**Comment by dme65** [comment]
> Can we do best value achieved rather than regret for consistency with the other tutorials?

We can do that! The main reason for using regret is to be able to put things on a log-scale, but we can skip that part.

> I wonder if it's worth demonstrating space complexities in addition to the time complexities (at least, if that isn't too hard to do). Seems that running out of memory is a recurring issue for users of TS.

Yeah, I will add some comments regarding memory usage. Anything with Cholesky is quadratic in the number of sample points while KeOps allows you to drop this to linear when used with CIQ.  RFFs will be linear without KeOps.

**Comment by Balandat** [comment]
This is great. A few additional comments: 

1. Maybe you can switch out items 2. and 3. in the introduction? 2 talks about CIQ before it is introduced in 3.

2. Do you mind adding the timings / memory footprint to this similar as in https://github.com/saitcakmak/gp-sampling/blob/master/notebooks/ciq_thompson_sampling.ipynb ? 

3. In the TS code, you can clean up repeated code by using an ExitStack:
```
    # Thompson sample
    with ExitStack() as es:
        if sampler == "cholesky":
            es.enter_context(gpts.max_cholesky_size(float("inf")))
        elif sampler == "ciq":
            es.enter_context(gpts.fast_computations(covar_root_decomposition=True))
            es.enter_context(gpts.max_cholesky_size(0))
            es.enter_context(gpts.ciq_samples(True))
            es.enter_context(gpts.minres_tolerance(5e-3))
            es.enter_context(gpts.num_contour_quadrature(15))
        elif sampler == "rff":
            es.enter_context(gpts.fast_computations(covar_root_decomposition=True))
    
        thompson_sampling = MaxPosteriorSampling(model=model, replacement=False)
        X_next = thompson_sampling(X_cand, num_samples=batch_size)
```

**Comment by dme65** [comment]
> This is great. A few additional comments:
> 
> 1. Maybe you can switch out items 2. and 3. in the introduction? 2 talks about CIQ before it is introduced in 3.
> 2. Do you mind adding the timings / memory footprint to this similar as in https://github.com/saitcakmak/gp-sampling/blob/master/notebooks/ciq_thompson_sampling.ipynb ?
> 3. In the TS code, you can clean up repeated code by using an ExitStack:
> 
> ```
>     # Thompson sample
>     with ExitStack() as es:
>         if sampler == "cholesky":
>             es.enter_context(gpts.max_cholesky_size(float("inf")))
>         elif sampler == "ciq":
>             es.enter_context(gpts.fast_computations(covar_root_decomposition=True))
>             es.enter_context(gpts.max_cholesky_size(0))
>             es.enter_context(gpts.ciq_samples(True))
>             es.enter_context(gpts.minres_tolerance(5e-3))
>             es.enter_context(gpts.num_contour_quadrature(15))
>         elif sampler == "rff":
>             es.enter_context(gpts.fast_computations(covar_root_decomposition=True))
>     
>         thompson_sampling = MaxPosteriorSampling(model=model, replacement=False)
>         X_next = thompson_sampling(X_cand, num_samples=batch_size)
> ```

Good comments! Let me know if the latest version looks good to you.

**Comment by saitcakmak** [comment]
> Do you mind adding the timings / memory footprint to this similar as in https://github.com/saitcakmak/gp-sampling/blob/master/notebooks/ciq_thompson_sampling.ipynb ?

FYI, I'll be sharing an updated version of this notebook with you guys soon. No need to look into it yet.

**Pull request closed without merging**